### DISCUSSION TITLE
Add VLES turbulence model in dafoam

### DISCUSSION THREAD

**Discussion by erkhans** [mdolab/dafoam#475] [Installation]
I have installed VLES in OpenFOAM. How do I transfer it into dafoam?

**Comment by friedenhe** [comment]
Please refer to the 2022 workshop slides on how to add new solvers/models to DAFoam. Also, is VLES an unsteady (time-dependent) model? DAFoam supports only steady-state optimization at this moment (it can run unsteady simulations but the unsteady adjoint is not implemented). We will hopefully be able to extend it to unsteady flow, maybe sometime next year.

**Reply by erkhans** [reply]
Thank you. Can I just add the turbulence model into the OpenFOAM-v1812 in dafoam folder? And indicate to use that model in case setup file "turbulenceProperties"? Will that work?

**Reply by friedenhe** [reply]
No, unfortunately, you need to also add the VLES model to src/adjoint/DAModel/DATurbulenceModel.

**Reply by erkhans** [reply]
I followed the steps shown in the slides, but when I recompile the dafoam the error shown in the figure appears. Adding a new model is shortly explained. Also, the video of the workshop does not include the part where the new model is added. Could you please give some suggestions? Thank you.
![error](https://github.com/mdolab/dafoam/assets/114105732/8194543d-e1eb-4a95-bf54-7154a65f7ec5)

**Reply by friedenhe** [reply]
it seems that the error is not related to the new vles model you added. can you try to compile the code without adding vles?

**Reply by erkhans** [reply]
So, when I recompile dafoam without adding anything, this is what I get. I can't understand what is wrong with it. 
![err](https://github.com/mdolab/dafoam/assets/114105732/b3ed51f4-321d-465e-80c4-156a8c97b431)
![err2](https://github.com/mdolab/dafoam/assets/114105732/09da4501-fe7a-40e8-85b8-3e11711440d2)

**Reply by friedenhe** [reply]
Are you sure you followed the exact steps to compile DAFoam's dependencies and DAFoam? What OS version do you use?

**Reply by erkhans** [reply]
Yes. I am using Ubuntu 20.04.6 LTS. Could please try to run ./Allmake command in your dafoam folder and see if it gives the same error?

**Comment by friedenhe** [comment]
We can compile the dafoam repo without a problem by running Allmake. From the first error you shared, it seems that your petsc installation was not correct. From the 2nd error, it seems that your folder structure in dafoam/OpenFOAM was not correct. Anyway, I suggest you start from scratch and follow the exact steps from the installation page.

**Reply by erkhans** [reply]
Thank you for your reply. I fixed this issue.  Now, I am getting this error shown in the screenshot when compiling. Do you have any suggestions on this?
![error3](https://github.com/mdolab/dafoam/assets/114105732/baa90a3b-c26c-4b9b-b6f0-916bcc141979)

**Comment by friedenhe** [comment]
It seems that you did not define some pure virtual functions in your new mykOmegaSST model.

**Reply by erkhans** [reply]
Thanks for your suggestions. I finally compiled it. But the dafoam cannot see the added model when running the test case (NACA airfoil). What other actions does it require to use the added model in simulations?

**Comment by friedenhe** [comment]
> But the dafoam cannot see the added model

Can you share the error log for this?

**Reply by erkhans** [reply]
Here 
<img width="740" alt="1" src="https://github.com/mdolab/dafoam/assets/114105732/b7777af5-cff0-4d74-9fdf-d6287c2a7093">

**Reply by erkhans** [reply]
[logOpt.txt](https://github.com/mdolab/dafoam/files/13173923/logOpt.txt)

**Comment by friedenhe** [comment]
Check if you have set `TypeName("mykOmegaSST")` in mykOmegaSST.H

**Reply by erkhans** [reply]
Yes, I have set it.

**Comment by friedenhe** [comment]
I suggest you create a fork of the DAFoam repo and push your development to your fork. We can then take a look at the code.

**Reply by erkhans** [reply]
Hi, I was able to compile the dafoam with the added model. Now, it can see the mykOmegaSST model but asks for the DATublenceModel. You can see in the figure.
<img width="447" alt="mykOmegaSST" src="https://github.com/mdolab/dafoam/assets/114105732/baf60246-78a2-4b63-8295-2f6ad9a1132c">

When I add the DATurbulenceModel, indicating the typename "mykOmegaSST", dafoam again cannot see the new model, shown in the figures below. If I use other name than "mykOmegaSST", it appears in the list of valid DATurbulenceModel types. So, the problem is I cannot use the same name when they have to be the same. Could you please help with this?

<img width="169" alt="typename" src="https://github.com/mdolab/dafoam/assets/114105732/e739610d-c788-4b34-b0af-5ccd7f5a2102">

<img width="322" alt="err" src="https://github.com/mdolab/dafoam/assets/114105732/fec38e02-4d79-495c-aeba-3c8d851f4922">

**Comment by friedenhe** [comment]
Can you upload the full log file? I suspect you compiled the original DAFoam version but did not compile the AD version. Note that both versions need to load your new turbulence model mykOmegaSST.

**Reply by erkhans** [reply]
[logfile.txt](https://github.com/mdolab/dafoam/files/13624185/logfile.txt)

Yes, I compiled the original DAFoam version. Could you please explain how to compile the AD version with the added model? Thank you.

**Comment by friedenhe** [comment]
Check here: https://dafoam.github.io/mydoc_installation_source.html#dafoam

**Comment by uttamcadambi07** [comment]
Hi @erkhans how did you resolve the problem? I am still facing the issue where if i add a new turbulence model (say kEpsilonRef), then I get an error that it is an unknown RAS model

**Reply by erkhans** [reply]
Hi, you also need to compile AD version. Try to compile using the lines below in dafoam/repos folder:
cd dafoam-* && \
. $HOME/dafoam/loadDAFoam.sh && \
export DAFOAM_NO_WARNINGS=1 && \
./Allmake && \
source $HOME/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/etc/bashrc && \
./Allclean && ./Allmake && \
source $HOME/dafoam/OpenFOAM/OpenFOAM-v1812-ADF/etc/bashrc && \
./Allclean && ./Allmake
